packages/frameworks-nextjs/src/index.tsx

```diff
@@ -0,0 +1,338 @@
+   * :::info
+   * Adding `auth` to your Middleware is optional, but recommended to keep the user session alive.
+   * :::
+   *
+   * Authentication is done by the {@link NextAuthConfig.callbacks|callbacks.authorized} callback.
+   * @example
+   * ```ts title="middleware.ts"
+   * export { auth as middleware } from "./auth"
+   * ```
+   *
+   * Alternatively you can wrap your own middleware with `auth`, where `req` is extended with `auth`:
+   * @example
+   * ```ts title="middleware.ts"
+   * import { auth } from "./auth"
+   * export default auth((req) => {
+   *   // req.auth
+   * })
+   * ```
+   *
+   * ```ts
+   * // Optionally, don't invoke Middleware on some paths
+   * // Read more: https://nextjs.org/docs/app/building-your-application/routing/middleware#matcher
+   * export const config = {
+   *   matcher: ["/((?!api|_next/static|_next/image|favicon.ico).*)"],
+   * }
+   * ```
+   *
+   * #### In Server Components
+   *
+   * @example
+   * ```ts title="app/page.ts"
+   * import { auth } from "../auth"
+   *
+   * export default async function Page() {
+   *   const { user } = await auth()
+   *   return <p>Hello {user?.name}</p>
+   * }
+   * ```
+   *
+   * #### In Route Handlers
+   * @example
+   * ```ts title="app/api/route.ts"
+   * import { auth } from "../../auth"
+   *
+   * export const POST = auth((req) => {
+   *   // req.auth
+   * })
+   * ```
+   *
+   * #### In Edge API Routes
+   *
+   * @example
+   * ```ts title="pages/api/protected.ts"
+   * import { auth } from "../../auth"
+   *
+   * export default auth((req) => {
+   *   // req.auth
+   * })
+   *
+   * export const config = { runtime: "edge" }
+   * ```
+   *
+   * #### In API Routes
+   *
+   * @example
+   * ```ts title="pages/api/protected.ts"
+   * import { auth } from "../auth"
+   * import type { NextApiRequest, NextApiResponse } from "next"
+   *
+   * export default async (req: NextApiRequest, res: NextApiResponse) => {
+   *   const session = await auth(req, res)
+   *   if (session) {
+   *     // Do something with the session
+   *     return res.json("This is protected content.")
+   *   }
+   *   res.status(401).json("You must be signed in.")
+   * }
+   * ```
+   *
+   * #### In `getServerSideProps`
+   *
+   * @example
+   * ```ts title="pages/protected-ssr.ts"
+   * import { auth } from "../auth"
+   * //...
+   * export const getServerSideProps: GetServerSideProps = async (context) => {
+   *   const session = await auth(context)
+   *
+   *   if (session) {
+   *     // Do something with the session
+   *     return { props: { session, content: (await res.json()).content } }
+   *   }
+   *
+   *   return { props: {} }
+   * }
+   * ```
+   */
+  auth: ((...args: [NextApiRequest, NextApiResponse]) => Promise<Session>) &
+    ((...args: []) => Promise<Session>) &
+    ((...args: [GetServerSidePropsContext]) => Promise<Session>) &
+    ((
+      ...args: [(req: NextAuthRequest) => ReturnType<AppRouteHandlerFn>]
+    ) => AppRouteHandlerFn)
```

Comment:
Was just trying this out locally, and small thing here - it seems that the `auth` response can be undefined, so `Promise<Session | undefined>` might be more accurate.